core/src/main/java/com/linecorp/armeria/common/DefaultFlagsProvider.java

```diff
@@ -0,0 +1,377 @@
+    static final long DEFAULT_REQUEST_TIMEOUT_MILLIS = 10 * 1000; // 10 seconds
+    static final int DEFAULT_HTTP1_MAX_INITIAL_LINE_LENGTH = 4096; // from Netty
+    static final int DEFAULT_HTTP1_MAX_HEADER_SIZE = 8192; // from Netty
+    static final int DEFAULT_HTTP1_MAX_CHUNK_SIZE = 8192; // from Netty
+    static final long DEFAULT_PING_INTERVAL_MILLIS = 0; // Disabled
+    static final int DEFAULT_MAX_SERVER_NUM_REQUESTS_PER_CONNECTION = 0; // Disabled
+    static final int DEFAULT_MAX_CLIENT_NUM_REQUESTS_PER_CONNECTION = 0; // Disabled
+    static final long DEFAULT_MAX_SERVER_CONNECTION_AGE_MILLIS = 0; // Disabled
+    static final long DEFAULT_MAX_CLIENT_CONNECTION_AGE_MILLIS = 0; // Disabled
+    static final long DEFAULT_SERVER_CONNECTION_DRAIN_DURATION_MICROS = 1000000;
+    static final int DEFAULT_HTTP2_INITIAL_CONNECTION_WINDOW_SIZE = 1024 * 1024; // 1MiB
+    static final int DEFAULT_HTTP2_INITIAL_STREAM_WINDOW_SIZE = 1024 * 1024; // 1MiB
+    static final int DEFAULT_HTTP2_MAX_FRAME_SIZE = 16384; // From HTTP/2 specification
+
+    // Can't use 0xFFFFFFFFL because some implementations use a signed 32-bit integer to store HTTP/2 SETTINGS
+    // parameter values, thus anything greater than 0x7FFFFFFF will break them or make them unhappy.
+    static final long DEFAULT_HTTP2_MAX_STREAMS_PER_CONNECTION = Integer.MAX_VALUE;
+    static final long DEFAULT_HTTP2_MAX_HEADER_LIST_SIZE = 8192L; // from Netty default maxHeaderSize
+    static final String DEFAULT_BACKOFF_SPEC = "exponential=200:10000,jitter=0.2";
+    static final int DEFAULT_MAX_TOTAL_ATTEMPTS = 10;
+    static final String ROUTE_CACHE_SPEC = "maximumSize=4096";
+    static final String ROUTE_DECORATOR_CACHE_SPEC = "maximumSize=4096";
+    static final String PARSED_PATH_CACHE_SPEC = "maximumSize=4096";
+    static final String HEADER_VALUE_CACHE_SPEC = "maximumSize=4096";
+    static final String CACHED_HEADERS = ":authority,:scheme,:method,accept-encoding,content-type";
+    static final String FILE_SERVICE_CACHE_SPEC = "maximumSize=1024";
+    static final String DNS_CACHE_SPEC = "maximumSize=4096";
+
+    private DefaultFlagsProvider() {}
+
+    @Override
+    public int priority() {
+        return Integer.MIN_VALUE;
+    }
+
+    @Override
+    public Sampler<Class<? extends Throwable>> verboseExceptionSampler() {
+        return VERBOSE_EXCEPTION_SAMPLER;
+    }
+
+    @Override
+    public Boolean verboseSocketExceptions() {
+        return false;
+    }
+
+    @Override
+    public Boolean verboseResponses() {
+        return false;
+    }
+
+    @Override
+    public RequestContextStorageProvider requestContextStorageProvider() {
+        final List<RequestContextStorageProvider> providers = FlagsUtil.getRequestContextStorageProviders();
+        if (providers.isEmpty()) {
+            return RequestContextStorage::threadLocal;
+        }
+        if (providers.size() > 1) {
+            throw new IllegalStateException(
```

Comment:
1) Multiple `RequestContextStorageProvider`s are already verified by `SystemPropertyFlagsProvider` and 
2) We assumed that `DefaultFlagsProvder` does not throw an exception.
https://github.com/line/armeria/blob/4ef61f1ef81f8e0d62ed016cf58ad05b3dae4ce8/core/src/main/java/com/linecorp/armeria/common/Flags.java#L1373-L1374
Therefore, why don't we always return `RequestContextStorage::threadLocal`?
We reach here only when `RequestContextStorageProvider` is misconfigured.

Reply:
Agreed. We could simplify this method body into just `return RequestContextStorage::threadLocal;` given that `SystemPropertyFlagsProvider` gives highest priority on SPI-based storage provider loading.

We might want to remove the SPI-based storage provider in 2.0 because a user can just implement a `FlagsProvider` instead after this PR is merged.

Reply:
Alternatively, how about making the following changes?

- Change `SystemPropertyFlagsProvider` return non-null only when there are one or more storage providers AND a user specified a class name in the system property, so its behavior adheres to its name (i.e. has effect only when there's a system property); and
- Change `DefaultFlagsProvider` returns the first storage provider OR the thread local one.

By doing this, we can allow a user's `FlagsProvider` implementation to override the storage provider conveniently. Currently, a user cannot override the storage provider using a `FlagsProvider` when there's a `FlagsProvider` loaded via SPI.

Reply:
Related: https://github.com/line/armeria/issues/4211

Reply:
That makes sense. I'll update the implementation. Thanks for advise.